src/extra/vui/vui.cpp

```diff
@@ -0,0 +1,243 @@
+#include "gfx/egui/egui_types.hpp"
+#include "gfx/linalg.hpp"
+#include "linalg_shim.hpp"
+#include "object_var_util.hpp"
+#include "params.hpp"
+#include "../egui/context.hpp"
+#include <gfx/renderer.hpp>
+#include <gfx/gizmos/gizmos.hpp>
+#include <input/input.hpp>
+
+namespace shards::VUI {
+
+#define VUI_PANEL_SHARD_NAME "VUI.Panel"
+
+struct VUIPanelShard;
+struct VUIContextShard {
+  VUIContext _vuiContext{};
+  SHVar *_vuiContextVar{};
+
+  Inputs::RequiredInputContext _inputContext;
+  gfx::RequiredGraphicsContext _graphicsContext;
+
+  ExposedInfo _exposedVariables;
+
+  std::vector<VUIPanelShard *> _panels;
+
+  input::InputBuffer _inputBuffer;
+
+  std::shared_ptr<gfx::GizmoRenderer> _debugRenderer;
+
+  PARAM_PARAMVAR(_queue, "Queue", "The draw queue to insert draw commands into.", {Type::VariableOf(gfx::Types::DrawQueue)});
+  PARAM_PARAMVAR(_view, "View", "The view that is being used to render.", {Type::VariableOf(gfx::Types::View)});
+  PARAM(ShardsVar, _contents, "Contents", "The list of UI panels to render.", {CoreInfo::ShardsOrNone});
+  PARAM_VAR(_scale, "Scale", "The scale of how many UI units per world unit.", {CoreInfo::FloatType});
+  PARAM_VAR(_debug, "Debug", "Render debug.", {CoreInfo::BoolType});
+  PARAM_IMPL(VUIContextShard, PARAM_IMPL_FOR(_queue), PARAM_IMPL_FOR(_view), PARAM_IMPL_FOR(_contents), PARAM_IMPL_FOR(_scale),
+             PARAM_IMPL_FOR(_debug));
+
+  VUIContextShard() {
+    _scale = Var(1000.0f);
+    _debug = Var(false);
+  }
+
+  static SHTypesInfo inputTypes() { return CoreInfo::NoneType; }
+  static SHTypesInfo outputTypes() { return CoreInfo::NoneType; }
+  static SHOptionalString help() {
+    return SHCCSTR("Creates a context for virtual UI panels to make sure input is correctly handled between them");
+  }
+
+  SHExposedTypesInfo requiredVariables() {
+    static auto e =
+        exposedTypesOf(decltype(_inputContext)::getExposedTypeInfo(), decltype(_graphicsContext)::getExposedTypeInfo());
+    return e;
+  }
+
+  void warmup(SHContext *context);
+
+  void cleanup() {
+    PARAM_CLEANUP()
+
+    _contents.cleanup();
+    _graphicsContext.cleanup();
+    _inputContext.cleanup();
+
+    if (_vuiContextVar) {
+      if (_vuiContextVar->refcount > 1) {
+        SHLOG_ERROR("VUI.Context: Found {} dangling reference(s) to {}", _vuiContextVar->refcount - 1, VUIContext::VariableName);
+      }
+      releaseVariable(_vuiContextVar);
+    }
+  }
+
+  SHTypeInfo compose(SHInstanceData &data) {
+    _exposedVariables = ExposedInfo(data.shared);
+    _exposedVariables.push_back(VUIContext::VariableInfo);
+    data.shared = SHExposedTypesInfo(_exposedVariables);
+
+    _contents.compose(data);
+
+    if (_queue->valueType == SHType::None)
+      throw ComposeError("Queue is required");
+
+    if (_view->valueType == SHType::None)
+      throw ComposeError("View is required");
+
+    return shards::CoreInfo::NoneType;
+  }
+
+  SHVar activate(SHContext *shContext, const SHVar &input) {
```

Comment:
Is there no way to pass the input of `VUI.Context` as input of each `VUI.Panel`?

Reply:
Yeah I can do that I think

Reply:
I tried to look at your solution for `UI.DockArea`, does it do that too?

Reply:
Yes.

It is given to the TabViewer (https://github.com/fragcolor-xyz/shards/blob/devel/rust/src/shards/gui/containers/docking.rs#L404). Then, in turn, passed to the contents children (https://github.com/fragcolor-xyz/shards/blob/devel/rust/src/shards/gui/containers/docking.rs#L453).